components/cdc/src/delegate.rs

```diff
@@ -671,6 +709,163 @@ impl Delegate {
         self.mark_failed();
         Err(Error::Request(store_err.into()))
     }
+
+    // Return Either::Right(Some(...)) if key was contained in the extra and can get it from short_value or default cf.
+    // Return Either::Right(None) if key was contained in the extra and cannot get it from short_value and default cf.
+    // Return Either::Left(key) if key wasn't contained in the extra, needs to be seeked from the engine.
+    fn old_value_from_extra(
+        &mut self,
+        row: &EventRow,
+        txn_extra: &mut TxnExtra,
+        reader: &mut OldValueReader<RegionSnapshot<RocksSnapshot>>,
+    ) -> Either<Key, Option<Value>> {
+        let old_values = txn_extra.mut_old_values();
+        let key = Key::from_raw(row.key.as_slice()).append_ts(row.start_ts.into());
+        if let Some((old_value, mutation_type)) = old_values.remove(&key) {
+            match mutation_type {
+                MutationType::Insert => assert!(old_value.is_none()),
+                MutationType::Put | MutationType::Delete => {
+                    if let Some(old_value) = old_value {
+                        let start_ts = old_value.start_ts;
+                        return Either::Right(old_value.short_value.or_else(|| {
+                            let prev_key = key.truncate_ts().unwrap().append_ts(start_ts);
+                            let mut opts = ReadOptions::new();
+                            opts.set_fill_cache(false);
+                            reader.get_value_default(&prev_key)
+                        }));
+                    }
+                }
+                _ => unreachable!(),
+            }
+        }
+        Either::Left(key)
+    }
+
+    // to_seek_old contains encoded keys.
+    fn seek_for_old_value(
+        &mut self,
+        to_seek_old: BTreeSet<Key>,
+        rows: &mut HashMap<Vec<u8>, EventRow>,
+        reader: &mut OldValueReader<RegionSnapshot<RocksSnapshot>>,
+    ) -> Result<()> {
+        for key in to_seek_old {
+            match reader.near_seek_old_value(&key)? {
+                Some(value) => {
+                    let raw_key = key.truncate_ts().unwrap().into_raw().unwrap();
+                    rows.get_mut(&raw_key).unwrap().old_value = value;
+                }
+                None => {
+                    self.mark_failed();
+                    let store_err = RaftStoreError::RegionNotFound(self.region_id);
```

Comment:
What about a key that was deleted by GC?

Reply:
In general, keys written after `checkpoint_ts` should not be GC.

Reply:
So it is still possible, right?

Reply:
Yes, also we need to check safepoint for `checkpoint_ts` from the request. See #8242
I suggest to address them together in another PR.

Reply:
Could it be resolved by using a snapshot that is taken at apply worker? The snapshot pins old values.

Reply:
Sure, but it's hard because of engine traits (cdc and apply worker do not use engine traits yet but observer have been abstracted in it.)

Reply:
And I'm working on applying engine traits to apply worker.